### PR TITLE
Copter: StandBy triggered by Pin

### PR DESCRIPTION
**Pull request by lthall**
This PR provides the ability to set the Stand By mode using an aux input from a carrier board.

### CODE REVIEW

File: ArduCopter/Parameters.cpp [hunk 1]
```diff
@@ -969,6 +969,14 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
     AP_SUBGROUPPTR(mode_zigzag_ptr, "ZIGZ_", 38, ParametersG2, ModeZigZag),
 #endif
 
+    // @Param: STANDBY_PIN
+    // @DisplayName: StandBy pin
+    // @Description: pin number to used to enable Stand By state. If set to -1 then it will not be used.
+    // @Values: -1:Disabled, 50:PX4 AUX1, 51:PX4 AUX2, 52:PX4 AUX3, 53:PX4 AUX4, 54:PX4 AUX5, 55:PX4 AUX6
```
Comment (khancyr) [inline ArduCopter/Parameters.cpp:975]: we need to remove px4 here as we don't px4 anymore!

File: ArduCopter/standby.cpp [hunk 2]
```diff
@@ -12,6 +12,20 @@
 //      and landing detection is disabled.
 void Copter::standby_update()
 {
+    if(g2.standby_pin > -1) {
```
Comment (khancyr) [inline ArduCopter/standby.cpp:15]: miss a space here : `if (`

**Comment by rmackay9** [comment]
I wonder if we should instead allow AP_Button to set an arbitrary flight mode?

**Comment by lthall** [comment]
> my only concern is that the pin will always overwrite the switch behavior

Yes, it would be I did not intend for them both to be used at the same time. The idea is to facilitate carrier board implementation of multi autopilot redundancy.



> I wonder if we should instead allow AP_Button to set an arbitrary flight mode?

This isn't a flight mode. I think I messed up a little calling it "stand by MODE" it is really a state not a flight mode. I suspect we may change the naming convention here a little. I do think this is a fundamental system state that we should support at that level. For example I do not think this should make use of LUA scripting.


Peter, I agree with what you are saying and would be very happy to get your help. Once I am out of the controllers I am less confident to say how to best implement things like this.

**Comment by peterbarker** [comment]
I've rejigged this as a separate object.

One of the big advantages I noted as I was doing this was leaving the determination of whether standby mode should be active or not to the Standby object itself.  This resolves @khancyr 's point about the aux pin always taking precedence, which may need some review.

**Comment by peterbarker** [comment]
Now a point on the new functionality.

There's no debounce protection on that auxillary pin.

@rmackay9 mentioned AP_Button - one of the advantages of AP_Button is that it does debouncing for you.

OTOH, it's a difficult user story to tell... first go and set up AP_Button like this, now go and set up standby mode like that, ...

I think we might need to have some debounce functionality available on the HAL somehow.

**Comment by peterbarker** [comment]
Oh.  Also, since we have parameters, perhaps we want to make the first one an "enable" parameter (`STANDBY_ENABLE`).  We should at least leave room for it.

**Comment by lthall** [comment]
Thanks Peter, I agree that an ENABLE parameter is a good suggestion.

Could we do an enable that also acts as an input type:
0 Disable
1 Channel X input
2 PIN input

or something like that?

**Comment by peterbarker** [comment]
On Tue, 30 Jun 2020, Leonard Hall wrote:

> Thanks Peter, I agree that an ENABLE parameter is a good suggestion.
> 
> Could we do an enable that also acts as an input type:
> 0 Disable
> 1 Channel X input
> 2 PIN input

I think we can do ENABLE parameters as a bitmask.....

But we could also have
STANDBY_ENABLE
STANDBY_OPTIONS
STANDBY_PIN


Where ENABLE remains a boolean
OPTIONS is a bitmask with initial entries USE_AUXPIN and USE_RCSWITCH

**Comment by lthall** [comment]
The other thing to keep in mind that this is not as critical to debounce the input. Debouncing is always a good idea though.

**Comment by lthall** [comment]
Whatever you think is most appropriate.  I do not have a deep enough appreciation of the memory issues to intelligently justify a position.

**Comment by peterbarker** [comment]
On Tue, 30 Jun 2020, Leonard Hall wrote:

> The other thing to keep in mind that this is not as critical to debounce the input. Debouncing is always a good idea though.

Depends on your electrics, doesn't it?  A transient at the wrong time 
might put you into standby for long enough to reset integrators and 
whatnot...

Admittedly, that's gotta be a pretty thin sort of a race.

**Comment by peterbarker** [comment]
I think this might now be better done using the new attach-RC-channel-function-to-AP_Button code.
 - parameters for inverting the signal
 - allows for PWM input rather than just logic signals
 - logging of the inputs
 - debouncing of logic signals

That only leaves the refactoring into a Standby object, which is probably not worth doing without other changes.

@lthall any reason to keep going with this PR?  I think the Button approach (I just tested it as working in master) is fine...

**Comment by lthall** [comment]
> I think this might now be better done using the new attach-RC-channel-function-to-AP_Button code.
> 
> * parameters for inverting the signal
> * allows for PWM input rather than just logic signals
> * logging of the inputs
> * debouncing of logic signals
> 
> That only leaves the refactoring into a Standby object, which is probably not worth doing without other changes.
> 
> @lthall any reason to keep going with this PR? I think the Button approach (I just tested it as working in master) is fine...

This functionality is used in integrated multi autopilot carrier boards so it will be primarily a digital logic driven feature. There are a number of companies using this in it's current form.

I am happy to do this any way that caters to this type of implementation but I think the functionality is important.

**Comment by peterbarker** [comment]
On Wed, 3 Feb 2021, Leonard Hall wrote:

> This functionality is used in integrated multi autopilot carrier boards so it will be primarily a digital logic driven feature. There are a number of
> companies using this in it's current form.
> 
> I am happy to do this any way that caters to this type of implementation but I think the functionality is important.

Cool.   Buttons are traditionally logic-level-triggered, but have the 
option of PWM to increase reliability.

I'll close this one - we can revisit if someone discovers that Button 
can't be used for this.

Ta!

**Comment by lthall** [comment]
Sorry Peter but closing this isn't an acceptable outcome unless you have another PR that provides this functionality already. If it can't be done this way then let's decide how it should be done and the PR can be fixed so it can be merged.

**Comment by lthall** [comment]
Just went over this with Randy. It appears that the buttons do precisely what this PR was attempting to implement. Sorry for the misunderstanding!

Closing again.

**Pull request closed without merging**